## Vast-tools Nextflow pipeline

### Options

```
    --reads     '/path/to/read/*_files.fastq'
    --db        '/path/to/VASTDB' 
    --groups    '/path/to/file/containing/insights/about/samples' (refer to the groups.txt file) 
    --groupA    'conditionA' 
    --groupB    'conditionB' 
    --gen_ref   'genome_reference' 
    --cores     'number_of_cores' 
    --output    '/path/to/output/directory'
    --expr      'yes/no' (differential expression analysis)
``` 

The databases are available on the [vast-tools github](https://github.com/vastgroup/vast-tools).

### Example

```
nextflow run /Xnfs/abc/nf_scratch/mmarchand/vast_projet/vast_tools.nf \
-profile psmn \
-c /Xnfs/abc/nf_scratch/mmarchand/vast_projet/nextflow.config \
--reads '/Xnfs/abc/nf_scratch/mmarchand/vast_projet/trimmed_fastq_test/*_R{1,2}_cutadapt_match.fastq.gz' \
--db '/Xnfs/abc/nf_scratch/mmarchand/vast_projet/VASTDB/' \
--groups '/Xnfs/abc/nf_scratch/mmarchand/vast_projet/groups.txt' \
--groupA 'siDDX5_17' \
--groupB 'siGL2' \
--gen_ref 'hg19' \
--cores '16' \
--output '/Xnfs/abc/nf_scratch/mmarchand/vast_projet/vast_out/'
--expr 'yes'
```

The groups file contains the names of the fastq files as first column (example: 5Y_siDDX5_17_B1 for 5Y_siDDX5_17_B1_R1_cutadapt_match.fastq.gz). The second column contains condition IDs (groupA and groupB).

The nextflow pipeline generate two final files, namely **final_INCLUSION_FILE.DIFF.txt** for the differential alternative splicing analysis, and the **DiffGE.tab** for the differential expression analysis (according to the --expr option).

## Command lines used to perform Differential Alternative Splicing (DAS) Analysis with Vast-tools

#### align (x number_of_replicates)

vast-tools align path/to/5Y_siDDX5_17_B1_R1_cutadapt_match.fastq.gz path/to/5Y_siDDX5_17_B1_R2_cutadapt_match.fastq.gz -sp genome_reference (hg19/hg38...) --dbDir path/to/VASTDB --cores 16 -o path/to/output/directory --expr (for DE analysis)

#### combine

vast-tools combine -o /path/to/output/directory (same that for align) -sp genome_reference --cores 16 --dbDir path/to/VASTDB -C (for DE analysis)

#### diff (for DAS analysis)

vast-tools diff -a 5Y_siDDX5_17_B1_R1_cutadapt_match,5Y_siDDX5_17_B2_R1_cutadapt_match -b 5Y_siGL2_B1_R1_cutadapt_match,5Y_siGL2_B2_R1_cutadapt_match --sampleNameA=siDDX5_17 --sampleNameB=siGL2 -i path/to/INCLUSION_LEVELS.tab 

*For paired-end fastq the paired replicates are merged under the name of the first file (R1).

**Check the name of the INCLUSION_LEVELS.tab (it depends on the genome reference and the number of replicates used). 


## Comparison between Vast-tools and rMATS outputs

The **process_rMATS_Vast_outputs.R** filters significant events and compare the results between the tools. It takes in input the path to raw rMATS results (JC) and the path to the vast_out directory (Vast-tools Nextflow pipeline results) in addition to the output path. The version of the human genome reference and the number of samples used are also required to determine the name of the inclusion table file generated by Vast-tools. Vast-tools can identify events without coordinated. These events cause issues for executing the **identify_common_events.py** script. Thus, there is an option to precise whether these lines must be conserved or not. These paths can be modified directly at the beginning of the R script.  

The **identify_common_events.py** script identifies the corresponding events between rMATS differential alternative splicing events and Vast-tools differential events or events annotated in VASTDB from the previous script outputs. It takes in input the output path of the **process_rMATS_Vast_output.R** script.

```
python3 identify_common_events.py ~/stage/results_hg19/
```

Finally, the **process_rMATS_Vast_outputs.R** use the python script outputs to quantify the number of common events and events detected by rMATS that could be detected by Vast-tools (annotated in VASTDB). The results are added to the complementary_insights.txt file. It takes in input the output path of the **process_rMATS_Vast_output.R** (can be modified at the beginning of the R script).

### Results

rMATS detects exon skipping (ES), mutually exclusive exons (MXE), retained intron (RI), alternative 5' and 3' splice sites (A5SS and A3SS) alternative splicing events. In other hand, Vast-tools detects and groups the ES, microexon skipping (MIC in event_type column) and MXE patterns into alternative exon skipping (EX) events. The RI, A5SS and A3SS patterns are also detected by the toolset. 
The tools are described in more detail in the M1-MEMOIRE-Marchand_Mehdi_2023-2024.pdf.

Four tables can be found in the result directory:  
  - rmats_summary.tab: number of differential alternative splicing (DAS) events and differentially alternatively spliced (DAS) genes according to the type of alternative splicing event.  
  - final_rmats_summary.tab: rmats_summary.tab without strict duplicates (events with strictly the same alternative exon coordinates).  
  - vast_summary.tab: number of DAS events and DAS genes according to the type of alternative splicing event.  
  - common_summary.tab: number of differentially alternatively spliced (DAS) genes identified by both tools.

These results are illustrated in the events_histo.pdf.

![DAS_plot](https://github.com/mehdiretif/M1_internship/blob/main/image/example_events_histo.png)


The common_events.txt file corresponds to the rMATS differential ES events identified by both tools. In other hand, the raw_common_events.txt file corresponds to the rMATS differential ES events annotated in VASTDB (according to the human genome version). **WARNING: some duplicates are present in these files due to rMATS results redundancy**. These files were obtained through the **identify_common_events.py** script that identify the Vast-tools events (differential ES events or ES annotated in VASTDB) that correspond to the rMATS differential ES events with a margin of alternative splicing exon coordinates of 5 nucleotides (the flanking exon coordinates are not taken into account). 
The number of differential ES events detected in common without duplicates are given in the complementary_insights.txt with some other insights. 

